primitives/block/src/macro_block.rs

```diff
@@ -197,12 +197,61 @@ pub struct MacroHeader {
     /// The root of the Merkle tree of the blockchain state. It just acts as a commitment to the
     /// state.
     pub state_root: Blake2bHash,
-    /// The root of the Merkle tree of the body. It just acts as a commitment to the body.
+    /// The hash of the body. It just acts as a commitment to the body.
     pub body_root: Blake2sHash,
     /// The root of the trie diff tree proof.
     pub diff_root: Blake2bHash,
     /// A merkle root over all of the transactions that happened in the current epoch.
     pub history_root: Blake2bHash,
+    /// Contains all the information regarding the next validator set, i.e. their validator
+    /// public key, their reward address and their assigned validator slots.
+    /// Is only Some when the macro block is an election block.
+    pub validators: Option<Validators>,
+    /// A bitset representing which validator slots will be prohibited from producing micro blocks or
+    /// proposing macro blocks in the batch following this macro block.
+    /// This set is needed for nodes that do not have the state as it is normally computed
+    /// inside the staking contract.
+    pub next_batch_initial_punished_set: BitSet,
+}
+
+impl MacroHeader {
+    /// Returns whether this macro block is an election block.
+    pub fn is_election(&self) -> bool {
+        Policy::is_election_block_at(self.block_number)
+    }
+
+    pub(crate) fn verify(&self) -> Result<(), BlockError> {
+        // Check that validators are only set on election blocks.
+        if self.is_election() != self.validators.is_some() {
+            return Err(BlockError::InvalidValidators);
+        }
+        Ok(())
+    }
+
+    fn legacy_body_root(&self) -> Blake2sHash {
+        let mut h = <Blake2sHash as HashOutput>::Builder::default();
+        self.legacy_body_serialize_content(&mut h).unwrap();
+        h.finish()
+    }
+
+    pub fn legacy_body_serialize_content<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        if let Some(ref validators) = self.validators {
+            let pk_tree_root = validators.hash::<Blake2sHash>();
+            pk_tree_root.serialize_to_writer(writer)?;
+        } else {
+            0u8.serialize_to_writer(writer)?;
+        }
+
+        let punished_set_hash = self
+            .next_batch_initial_punished_set
+            .serialize_to_vec()
+            .hash::<Blake2sHash>();
+        punished_set_hash.serialize_to_writer(writer)?;
+
+        self.body_root.serialize_to_writer(writer)?;
+
+        Ok(())
+    }
```

Comment:
How long will these two be necessary? I suspect forever because of the zkp setup and then I would maybe consider naming them as such, to not make it seem like they can be removed at some later time.